### PR TITLE
Fix bug when a token is given public only

### PR DESCRIPTION
**Pull request by lunny**

### CODE REVIEW

**Review by delvh — APPROVED** [review]

File: routers/api/v1/api.go [hunk 1]
```diff
@@ -398,6 +407,23 @@ func reqRepoBranchWriter(ctx *context.APIContext) {
...
+func checkPublicOnly(ctx *context.APIContext, unitType unit.Type) {
+	if true == ctx.Data["IsApiToken"] {
+		switch unitType {
+		case unit.TypeCode:
+			publicRepo, pubRepoExists := ctx.Data["ApiTokenScopePublicRepoOnly"]
+
+			if pubRepoExists && publicRepo.(bool) &&
+				ctx.Repo.Repository != nil && ctx.Repo.Repository.IsPrivate {
+				ctx.Error(http.StatusForbidden, "reqToken", "token scope is limited to public repos")
+				return
```
Comment (yp05327) [inline routers/api/v1/api.go]: What is the difference between this and `reqToken`
No matter whether the unit type is, we will always check it in `reqToken`
And all routers which contain all these modified functions will always call `reqToken` first.
So it seems that this change has no effect? 🤔

Comment (yp05327) [reply routers/api/v1/api.go]: I see. write permission always has `reqToken` but read permission may not have public only check.
e.g. 
image

But this fix seems not good enough. As in `reqToken` there's another public only check for organizations.
Maybe we can reuse these codes.

Comment (lunny) [reply routers/api/v1/api.go]: These code are outdated.

**Review by Zettat123 — APPROVED** [review]

File: routers/api/v1/api.go [hunk 2]
```diff
@@ -235,6 +235,62 @@ func reqPackageAccess(accessMode perm.AccessMode) func(ctx *context.APIContext)
...
+			}
+		case auth_model.ContainsCategory(requiredScopeCategories, auth_model.AccessTokenScopeCategoryOrganization):
+			if ctx.Org.Organization != nil && ctx.Org.Organization.Visibility != api.VisibleTypePublic {
+				ctx.Error(http.StatusForbidden, "reqToken", "token scope is limited to public orgs")
+				return
+			}
+			if ctx.ContextUser != nil && ctx.ContextUser.IsOrganization() && ctx.ContextUser.Visibility != api.VisibleTypePublic {
+				ctx.Error(http.StatusForbidden, "reqToken", "token scope is limited to public orgs")
+				return
+			}
```
Comment (yp05327) [inline routers/api/v1/api.go:269]: I think this is same. Only check ContextUser is enough.

File: routers/api/v1/api.go [hunk 3]
```diff
@@ -235,6 +235,62 @@ func reqPackageAccess(accessMode perm.AccessMode) func(ctx *context.APIContext)
...
+		case auth_model.ContainsCategory(requiredScopeCategories, auth_model.AccessTokenScopeCategoryUser):
+			if ctx.ContextUser != nil && !ctx.ContextUser.IsOrganization() && ctx.ContextUser.Visibility != api.VisibleTypePublic {
```
Comment (yp05327) [inline routers/api/v1/api.go]: Not sure whether !ctx.ContextUser.IsOrganization() is ok? As we have many user types.

File: routers/api/v1/api.go (hunk 2 above)
Comment (lunny) [reply routers/api/v1/api.go:269]: I haven't checked every place, so keep both will make it more safer and harmless.

File: routers/api/v1/api.go (hunk 3 above)
Comment (lunny) [reply routers/api/v1/api.go]: 046dec4

File: routers/api/v1/api.go [hunk 4]
```diff
@@ -295,25 +347,6 @@ func reqToken() func(ctx *context.APIContext) {
 			return
 		}
 
-		if true == ctx.Data["IsApiToken"] {
```
Comment (yp05327) [inline routers/api/v1/api.go:298]: I'm not sure whether it is safe to remove it.
How to approve it? Did you check all router paths?

Comment (lunny) [reply routers/api/v1/api.go:298]: Yes. I checked all routers and also add some tests for the newly added code. These code have been extracted to a new method checkTokenPublicOnly.

**Review by lafriks — APPROVED** [review]

**Pull request merged**